src/protocols/vnc/display.c

```diff
@@ -0,0 +1,207 @@
+
+    int dx, dy;
+
+    /* Cairo image buffer */
+    int stride;
+    unsigned char* buffer;
+    unsigned char* buffer_row_current;
+    cairo_surface_t* surface;
+
+    /* VNC framebuffer */
+    unsigned int bpp;
+    unsigned int fb_stride;
+    unsigned char* fb_row_current;
+
+    /* Ignore extra update if already handled by copyrect */
+    if (vnc_client->copy_rect_used) {
+        vnc_client->copy_rect_used = 0;
+        return;
+    }
+
+    /* Init Cairo buffer */
+    stride = cairo_format_stride_for_width(CAIRO_FORMAT_RGB24, w);
+    buffer = malloc(h*stride);
+    buffer_row_current = buffer;
+
+    bpp = client->format.bitsPerPixel/8;
+    fb_stride = bpp * client->width;
+    fb_row_current = client->frameBuffer + (y * fb_stride) + (x * bpp);
+
+    /* Copy image data from VNC client to PNG */
+    for (dy = y; dy<y+h; dy++) {
+
+        unsigned int*  buffer_current;
+        unsigned char* fb_current;
+        
+        /* Get current buffer row, advance to next */
+        buffer_current      = (unsigned int*) buffer_row_current;
+        buffer_row_current += stride;
+
+        /* Get current framebuffer row, advance to next */
+        fb_current      = fb_row_current;
+        fb_row_current += fb_stride;
+
+        for (dx = x; dx<x+w; dx++) {
+
+            unsigned char red, green, blue;
+            unsigned int v;
+
+            switch (bpp) {
+                case 4:
+                    v = *((uint32_t*)  fb_current);
+                    break;
+
+                case 2:
+                    v = *((uint16_t*) fb_current);
+                    break;
+
+                default:
+                    v = *((uint8_t*)  fb_current);
+            }
+
+            /* Translate value to RGB */
+            red   = (v >> client->format.redShift)   * 0x100 / (client->format.redMax  + 1);
+            green = (v >> client->format.greenShift) * 0x100 / (client->format.greenMax+ 1);
+            blue  = (v >> client->format.blueShift)  * 0x100 / (client->format.blueMax + 1);
+
+            /* Output RGB */
+            if (vnc_client->settings->swap_red_blue)
+                *(buffer_current++) = (blue << 16) | (green << 8) | red;
+            else
+                *(buffer_current++) = (red  << 16) | (green << 8) | blue;
+
+            fb_current += bpp;
+
+        }
+    }
+
+    /* For now, only use default layer */
+    surface = cairo_image_surface_create_for_data(buffer, CAIRO_FORMAT_RGB24, w, h, stride);
+
+    guac_common_surface_draw(vnc_client->display->default_surface,
+            x, y, surface);
+
+    /* Free surface */
+    cairo_surface_destroy(surface);
+    free(buffer);
+
+}
+
+void guac_vnc_copyrect(rfbClient* client, int src_x, int src_y, int w, int h, int dest_x, int dest_y) {
+
+    guac_client* gc = rfbClientGetClientData(client, GUAC_VNC_CLIENT_KEY);
+    guac_vnc_client* vnc_client = (guac_vnc_client*) gc->data;
+
+    /* For now, only use default layer */
```

Comment:
This is an ancient comment, but I'm not sure what the plan was which would involve anything *but* default... I'll fix this.